docs/patterns/outbox.md

```diff
@@ -0,0 +1,121 @@
+
+While this characteristic is great for ensuring no information is lost by the running system, it has an inherent drawback that any given message may be delivered to the processing code multiple times. In absence of code that guards against processing a message multiple times the data quickly becomes corrupted by updates applied multiple times.
+
+Historically this issue has been mitigated by distributed transcation technologies implementing Two-Phase Commit protocol (2PC) like Microsoft Distributed Transaction Coordinator. Such technologies allowed a single transaction to span both the messaging infrastructure and the database, ensuring that either:
+ - all the data is updated and the message is consumed
+ - no data is updated and the message is returned back to the queue
+
+Due to the issues with scalability and performance of the 2PC protocol, the distributed transaction technologies are no longer available for modern messaging technologies, such as Azure Service Bus and developers need to write code protecting against duplicate processing into their message handling logic. Adding such logic to a message handling code is referred to as making that code _idempotent_. 
+
+In mathematics, where the term originates, _idempotence_ is a property of fuctions. A given function _f_ is _idempotent_ if and only if _f(f(x)) = (x)_. In the cotext of the HTTP protocol, [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4), states, "A ... method is considered _idempotent_ if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request.". The latter is hepful in attempt to define _idempotence_ for the message handling problem:
+
+> An independent message handling code is one that produces the same side effects regardless of how many times a given message has been delivered
+
+The well-documented approach to the challenge of making message handling code _idempotent_ is to add a check, prior to executing the code, if the side effects have already been applied. This is illustrated by the following pseudocode:
+
+```c#
+if (!HasBeenProcessed(message))
+{
+    return;
+}
+Process(message);
+MarkProcessed(message);
+```
+
+It does not, however, take into account that a typical message handling code affects multiple external resources, i.e. a database (store or update some data) and a messaging infrastructure (send follow-up messages). In other words, it may happen that the first attempt at processing a message resulted in updating a database but failed to send out messages. The message is put back into the queue, picked up again and the second attempt manages to send the messages. 
```

Comment:
```suggestion
It does not, however, take into account that typical message handling code affects multiple external resources, i.e. a database (store or update some data) and messaging infrastructure (send follow-up messages). In other words, it may happen that the first attempt at processing a message resulted in updating the database but crashed before being able to send out messages. In that case, the message is "rolled back" to the queue, after which it will be picked up and attempted to be processed again. 
```